Aquarius/Src/Aquarius/Renderer/Shader.cpp

```diff
@@ -0,0 +1,177 @@
+            AQ_CORE_ERROR("Error: %v", error.what());
+        }
+
+        uint32_t vertexID;
+        uint32_t fragmentID;
+        vertexID = glCreateShader(GL_VERTEX_SHADER);
+        fragmentID = glCreateShader(GL_FRAGMENT_SHADER);
+
+        compile(vertexID, vertexCode.c_str(), ShaderType::VERTEX_SHADER);
+        compile(fragmentID, fragmentCode.c_str(), ShaderType::FRAGMENT_SHADER);
+        link(vertexID, fragmentID);
+
+        glDeleteShader(vertexID);
+        glDeleteShader(fragmentID);
+    }
+
+    Shader::~Shader()
+    {
+        deactivate();
+        glDeleteProgram(m_ID);
+    }
+
+    void Shader::activate()
+    {
+        glUseProgram(m_ID);
+    }
+
+    void Shader::deactivate()
+    {
+        glUseProgram(0);
+    }
+
+    void Shader::setBool(const std::string& name, bool value) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform1i(location, (int)value);
+    }
+
+    void Shader::setInt(const std::string& name, int value) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform1i(location, value);
+    }
+
+    void Shader::setInt2(const std::string& name, const glm::ivec2& values) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform2i(location, values.x, values.y);
+    }
+
+    void Shader::setInt3(const std::string& name, const glm::ivec3& values) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform3i(location, values.x, values.y, values.z);
+    }
+
+    void Shader::setInt4(const std::string& name, const glm::ivec4& values) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform4i(location, values.x, values.y, values.z, values.w);
+    }
+
+    void Shader::setFloat(const std::string& name, float value) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform1f(location, value);
+    }
+
+    void Shader::setFloat2(const std::string& name, const glm::vec2& values) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform2f(location, values.x, values.y);
+    }
+
+    void Shader::setFloat3(const std::string& name, const glm::vec3& values) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform3f(location, values.x, values.y, values.z);
+    }
+
+    void Shader::setFloat4(const std::string& name, const glm::vec4& values) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniform4i(location, values.x, values.y, values.z, values.w);
+    }
+
+    void Shader::setMat4(const std::string& name, bool transpose, const glm::mat4& value) const
+    {
+        int location = glGetUniformLocation(m_ID, name.c_str());
+        glUniformMatrix4fv(location, 1, transpose, glm::value_ptr(value));
+    }
```

Comment:
If only glfw decided to use C++/templates :(

Reply:
Hahaha